File: dnn-providers/hip-kernel-provider/rocKE/tests/instances/parity/pooling_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/pooling_emit.py -- Python reference emitter for the pooling2d
# instance parity harness. Selects one of the sampled configs by argv[1] (the
# config index), builds the Pooling2DSpec, builds the kernel via
# build_pooling2d and prints lower_kernel_to_llvm(arch='gfx950') to stdout so
# it can be byte-compared with the C emitter pooling_emit.c.
from rocke.instances.common.pooling import (
    Pooling2DSpec,
    PoolingProblem,
    build_pooling2d,
)
from _emit_common import run_emit


def _spec(idx: int) -> Pooling2DSpec:
    if idx == 0:
        return Pooling2DSpec(
            problem=PoolingProblem(
                N=1, H=28, W=28, C=64, Y=2, X=2, sH=2, sW=2, pH=0, pW=0, dH=1, dW=1
            ),
            dtype="f16",
            op="max",
            block_size=256,
            vec=1,
        )
    if idx == 1:
        return Pooling2DSpec(
            problem=PoolingProblem(
                N=2, H=56, W=56, C=128, Y=3, X=3, sH=1, sW=1, pH=1, pW=1, dH=1, dW=1
            ),
            dtype="f16",
            op="avg",
            block_size=256,
            vec=2,
        )
    if idx == 2:
        return Pooling2DSpec(
            problem=PoolingProblem(
                N=4, H=112, W=112, C=256, Y=7, X=7, sH=7, sW=7, pH=0, pW=0, dH=1, dW=1
            ),
            dtype="f16",
            op="sum",
            block_size=256,
            vec=4,
        )
    if idx == 3:
        return Pooling2DSpec(
            problem=PoolingProblem(
                N=1, H=224, W=224, C=64, Y=2, X=2, sH=2, sW=2, pH=0, pW=0, dH=1, dW=1
            ),
            dtype="bf16",
            op="max",
            block_size=256,
            vec=1,
        )
    if idx == 4:
        return Pooling2DSpec(
            problem=PoolingProblem(
                N=2, H=32, W=32, C=256, Y=3, X=3, sH=1, sW=1, pH=1, pW=1, dH=1, dW=1
            ),
            dtype="f16",
            op="avg",
            block_size=128,
            vec=8,
        )
    if idx == 5:
        return Pooling2DSpec(
            problem=PoolingProblem(
                N=1, H=64, W=64, C=128, Y=2, X=2, sH=2, sW=2, pH=0, pW=0, dH=1, dW=1
            ),
            dtype="bf16",
            op="sum",
            block_size=512,
            vec=4,
        )
    raise SystemExit(f"unknown config index {idx}")


def _build(spec, arch=None):
    return build_pooling2d(spec)


def main() -> int:
    return run_emit(
        _spec,
        _build,
        usage="usage: pooling_emit.py <config_index 0..5> [mode]\n",
    )


if __name__ == "__main__":
    raise SystemExit(main())

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/pooling_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/pooling_emit.c -- C-side emitter for the pooling2d instance
 * parity harness. Selects one of the sampled configs by argv[1] (the config
 * index), builds rocke_pooling2d_spec_t identically to the Python emitter
 * pooling_emit.py, inits a fresh IRBuilder with the spec kernel name, builds
 * into it via rocke_build_pooling2d (the C build entry), lowers via
 * rocke_lower_kernel_to_llvm (arch gfx950, flavor AUTO) and prints the .ll to
 * stdout so the two outputs can be byte-compared.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_pooling.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Fill `spec` for config index `idx`. Returns 0 on success, -1 if unknown. */
static int make_spec(int idx, rocke_pooling2d_spec_t* spec)
{
    *spec = rocke_pooling2d_spec_default();
    rocke_pooling_problem_t* p = &spec->problem;

    switch(idx)
    {
    case 0:
        p->N = 1;
        p->H = 28;
        p->W = 28;
        p->C = 64;
        p->Y = 2;
        p->X = 2;
        p->sH = 2;
        p->sW = 2;
        p->pH = 0;
        p->pW = 0;
        p->dH = 1;
        p->dW = 1;
        spec->dtype = "f16";
        spec->op = "max";
        spec->block_size = 256;
        spec->vec = 1;
        break;
    case 1:
        p->N = 2;
        p->H = 56;
        p->W = 56;
        p->C = 128;
        p->Y = 3;
        p->X = 3;
        p->sH = 1;
        p->sW = 1;
        p->pH = 1;
        p->pW = 1;
        p->dH = 1;
        p->dW = 1;
        spec->dtype = "f16";
        spec->op = "avg";
        spec->block_size = 256;
        spec->vec = 2;
        break;
    case 2:
        p->N = 4;
        p->H = 112;
        p->W = 112;
        p->C = 256;
        p->Y = 7;
        p->X = 7;
        p->sH = 7;
        p->sW = 7;
        p->pH = 0;
        p->pW = 0;
        p->dH = 1;
        p->dW = 1;
        spec->dtype = "f16";
        spec->op = "sum";
        spec->block_size = 256;
        spec->vec = 4;
        break;
    case 3:
        p->N = 1;
        p->H = 224;
        p->W = 224;
        p->C = 64;
        p->Y = 2;
        p->X = 2;
        p->sH = 2;
        p->sW = 2;
        p->pH = 0;
        p->pW = 0;
        p->dH = 1;
        p->dW = 1;
        spec->dtype = "bf16";
        spec->op = "max";
        spec->block_size = 256;
        spec->vec = 1;
        break;
    case 4:
        p->N = 2;
        p->H = 32;
        p->W = 32;
        p->C = 256;
        p->Y = 3;
        p->X = 3;
        p->sH = 1;
        p->sW = 1;
        p->pH = 1;
        p->pW = 1;
        p->dH = 1;
        p->dW = 1;
        spec->dtype = "f16";
        spec->op = "avg";
        spec->block_size = 128;
        spec->vec = 8;
        break;
    case 5:
        p->N = 1;
        p->H = 64;
        p->W = 64;
        p->C = 128;
        p->Y = 2;
        p->X = 2;
        p->sH = 2;
        p->sW = 2;
        p->pH = 0;
        p->pW = 0;
        p->dH = 1;
        p->dW = 1;
        spec->dtype = "bf16";
        spec->op = "sum";
        spec->block_size = 512;
        spec->vec = 4;
        break;
    default:
        return -1;
    }
    return 0;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index 0..5> [mode]\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    rocke_pooling2d_spec_t spec;
    if(make_spec(idx, &spec) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    const char* arch = "gfx950";

    /* Validate the spec (mirrors is_valid_spec). */
    char reason[ROCKE_ERR_MSG_CAP];
    reason[0] = 0;
    if(!rocke_pooling2d_is_valid_spec(&spec, arch, reason, sizeof reason))
    {
        fprintf(stderr, "invalid spec: %s\n", reason);
        return 1;
    }

    /* Init IRBuilder with spec.kernel_name() then build into it via the
     * task-specified entry rocke_build_pooling2d(b, &spec, arch). */
    char name[ROCKE_ERR_MSG_CAP];
    if(rocke_pooling2d_kernel_name(&spec, name, sizeof name) != ROCKE_OK)
    {
        fprintf(stderr, "kernel_name failed\n");
        return 1;
    }
    rocke_ir_builder_t b;
    if(rocke_ir_builder_init(&b, name) != ROCKE_OK)
    {
        fprintf(stderr, "builder init failed\n");
        return 1;
    }
    rocke_kernel_def_t* kernel = rocke_build_pooling2d(&b, &spec, arch);
    if(!kernel)
    {
        fprintf(stderr, "build failed: %s\n", b.err);
        rocke_ir_builder_free(&b);
        return 1;
    }

    if(strcmp(mode, "ll") == 0)
    {
        char* llvm_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, arch, &llvm_text);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* t = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &t);
        if(st != ROCKE_OK || !t)
        {
            fprintf(stderr, "serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(t, stdout);
        free(t);
    }
    else if(strcmp(mode, "verify") == 0)
    {
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s = rocke_diag_to_string(&d[i]);
            if(s)
            {
                puts(s);
                free(s);
            }
        }
        rocke_diags_free(d, n);
    }
    else
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        rocke_ir_builder_free(&b);
        return 2;
    }
    rocke_ir_builder_free(&b);
    return 0;
}
